pkg/controller/status/controller.go

```diff
@@ -0,0 +1,424 @@
+}
+
+// NewController creates a statusMessage controller, responsible for monitoring the operators statusMessage and updating its cluster statusMessage accordingly.
+func NewController(client configv1client.ConfigV1Interface, configurator Configurator, namespace string) *Controller {
+	c := &Controller{
+		name:         "insights",
+		statusCh:     make(chan struct{}, 1),
+		configurator: configurator,
+		client:       client,
+		namespace:    namespace,
+	}
+	return c
+}
+
+func (c *Controller) triggerStatusUpdate() {
+	select {
+	case c.statusCh <- struct{}{}:
+	default:
+	}
+}
+
+func (c *Controller) controllerStartTime() time.Time {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+	if c.start.IsZero() {
+		c.start = time.Now()
+	}
+	return c.start
+}
+
+// LastReportedTime provides the last reported time in a thread-safe way.
+func (c *Controller) LastReportedTime() time.Time {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+	return c.reported.LastReportTime.Time
+}
+
+// SetLastReportedTime sets the last reported time in a thread-safe way.
+func (c *Controller) SetLastReportedTime(at time.Time) {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+	if c.reported.LastReportTime.IsZero() {
+		klog.V(2).Infof("Initializing last reported time to %s", at.UTC().Format(time.RFC3339))
+	}
+	c.reported.LastReportTime.Time = at
+	c.triggerStatusUpdate()
+}
+
+// AddSources adds sources in a thread-safe way.
+// A source is used to monitor parts of the operator.
+func (c *Controller) AddSources(sources ...controllerstatus.Interface) {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+	c.sources = append(c.sources, sources...)
+}
+
+// Sources provides the sources in a thread-safe way.
+// A source is used to monitor parts of the operator.
+func (c *Controller) Sources() []controllerstatus.Interface {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+	return c.sources
+}
+
+func (c *Controller) merge(clusterOperator *configv1.ClusterOperator) *configv1.ClusterOperator {
+	// prime the object if it does not exist
+	if clusterOperator == nil {
+		clusterOperator = newClusterOperator(c.name, nil)
+	}
+
+	ctrlStatus := newControllerStatus()
+	// calculate the current controller state
+	allReady, lastTransition := c.currentControllerStatus(ctrlStatus)
+
+	clusterOperator = clusterOperator.DeepCopy()
+	now := time.Now()
+	if len(c.namespace) > 0 {
+		clusterOperator.Status.RelatedObjects = relatedObjects(c.namespace)
+	}
+
+	isInitializing := !allReady && now.Sub(c.controllerStartTime()) < 3*time.Minute
+
+	// cluster operator conditions
+	cs := newConditions(&clusterOperator.Status, metav1.Time{Time: now})
+	updateDisabledAndFailingConditions(cs, ctrlStatus, isInitializing, lastTransition)
+
+	// once the operator is running it is always considered available
+	cs.setCondition(configv1.OperatorAvailable, configv1.ConditionTrue, "AsExpected", "", metav1.Now())
+
+	updateProcessingConditionWithSummary(cs, ctrlStatus, isInitializing, lastTransition)
+
+	klog.V(4).Infof("The operator is healthy")
```

Comment:
I think we shouldn't always print this to log. An example from my log shows it:
```
I0923 10:19:56.005695       1 controller.go:203] Number of lastTransition upload failures 5 exceeded than threshold 5. Marking as degraded.
I0923 10:19:56.005733       1 controller.go:336] The operator has some internal errors: Unable to report: gateway server reported unexpected error code: 404 (request=259e010ba9a44a07ae7ab367ab2d2b0e): 404 page not found
I0923 10:19:56.005742       1 controller.go:371] The operator has some internal errors: 
I0923 10:19:56.005747       1 controller.go:152] The operator is healthy
````
Also the line with `The operator has some internal errors: ` is missing something. I guess the actual error.

Reply:
Ohh I realized it's twice there....we can keep only the first one then :)

Reply:
OK. But now I'm wondering why the second one doesn't have the reason. :thinking: I guess that I'll take a look just to make sure.

Reply:
This discussion was resolved, but the problem was not fully resolved. I am still seeing:
```
I1005 11:36:23.480363       1 controller.go:374] The operator has some internal errors: Unable to report: gateway server reported unexpected error code: 404 (request=8d7a0361f47641168e5d8da1b1598e0c): 404 page not found
I1005 11:36:23.480373       1 controller.go:157] The operator is healthy
```
I believe this was not case before to have one message saying there's some error and second message saying the operator is healthy.

Reply:
^ That's indeed pretty weird and doesn't seem right, but I feel like IO sometimes really reports `The operator is healthy` even after printing an error (not necessarily from this code, but anything that happens before the message is logged).

Reply:
You are right. I made a mistake here. Already fixed, let me just test it and check if I'm not forgetting anything. I got a little bit confused about the way that the `switch` should work, I guess...

Reply:
Now, it should be fine.